File: actions_py/actions_py/count_until_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from my_robot_interfaces.action import CountUntil
 
 
class CountUntilClient(Node): # count_until_client
    def __init__(self):
        super().__init__("count_until_client") # count_until_client
        self.count_until_client_ = ActionClient(self,CountUntil,"count_until")

    def send_goal(self,target_number,period):
        # wait for the server
        while not self.count_until_client_.wait_for_server(1.0):
            self.get_logger().info("Client is waiting for the server to start.")
        
        # goal
        goal = CountUntil.Goal()
        goal.target_number = target_number
        goal.period = period
 
        # sending goal
        self.get_logger().info("Sending Goal")
        self.count_until_client_.send_goal_async(goal,feedback_callback=self.goal_feedback_callback).add_done_callback(self.goal_response_callback)
        #self.timer_ = self.create_timer(2.0,self.cancel_goal)

    def cancel_goal(self):
        self.get_logger().info("Send a cancel Request.")
        self.goal_handle_.cancel_goal_async()
        self.timer_.cancel()

    def goal_response_callback(self,future):
        self.goal_handle_:ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal got accepted.")
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)
        else:
             self.get_logger().warn("Goal got rejected")


    def goal_result_callback(self,future):
        status = future.result().status
        result = future.result().result
        
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Success")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted") 
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().error("Canceled") 

        self.get_logger().info("Result: " + str(result.reached_number))

    def goal_feedback_callback(self,feedback_msg):
        number = feedback_msg.feedback.current_number
        self.get_logger().info("Got Feedback: " + str(number))

def main(args=None):
    rclpy.init(args=args)
    node = CountUntilClient() # count_until_client
    node.send_goal(10,1.0)
    rclpy.spin(node)
    rclpy.shutdown()
  
 
if __name__ == "__main__":
    main()

File: actions_cpp/src/count_until_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"
 
using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ClientGoalHandle<CountUntil>;
using namespace std::placeholders;

class CountUntilClientNode : public rclcpp::Node 
{
public:
    CountUntilClientNode() : Node("count_until_client") 
    {
        count_until_client_ = rclcpp_action::create_client<CountUntil>(this,"count_until");
    }
    
    void send_goal(int target_number, double period){
        // Wait for the action server
        count_until_client_->wait_for_action_server();

        // create a goal
        auto goal = CountUntil::Goal();
        goal.target_number = target_number;
        goal.period = period;

        // options
        auto options = rclcpp_action::Client<CountUntil>::SendGoalOptions();
        options.goal_response_callback = std::bind(&CountUntilClientNode::goal_response_callback,this,_1);
        options.result_callback = std::bind(&CountUntilClientNode::goal_result_callback,this,_1);
        options.feedback_callback = std::bind(&CountUntilClientNode::goal_feedback_callback,this,_1,_2);

        // send a goal
        RCLCPP_INFO(this->get_logger(),"Sending a Goal");
        count_until_client_->async_send_goal(goal,options);

        //timer_ = this->create_wall_timer(std::chrono::seconds(2),std::bind(&CountUntilClientNode::timer_callback,this));
    }
    
private:

    void timer_callback(){
        RCLCPP_INFO(this->get_logger(),"Cancel the Goal");
        count_until_client_->async_cancel_goal(goal_handle_);
        timer_->cancel();
    }

    void goal_feedback_callback(const CountUntilGoalHandle::SharedPtr &goal_handle, const std::shared_ptr<const CountUntil::Feedback> feedback){
        (void)goal_handle;
        int number = feedback->current_number;
        RCLCPP_INFO(this->get_logger(),"Got Feedback: %d",number);
    }

    void goal_response_callback(const CountUntilGoalHandle::SharedPtr &goal_handle){
        if(!goal_handle){
            RCLCPP_INFO(this->get_logger(),"Goal got Rejected");
        }else{
            this->goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(),"Goal got Accepted");
        }
    }


    void goal_result_callback(const CountUntilGoalHandle::WrappedResult &result)
    {   
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED){
            RCLCPP_INFO(this->get_logger(),"Succeeded!");            
        }else if (status == rclcpp_action::ResultCode::ABORTED){
            RCLCPP_ERROR(this->get_logger(),"Aborted!");            
        }else if (status == rclcpp_action::ResultCode::CANCELED){
            RCLCPP_ERROR(this->get_logger(),"Canceled!");            
        }  
        int reached_number = result.result->reached_number;
        RCLCPP_INFO(this->get_logger(),"Result: %d",reached_number);
    }
    rclcpp_action::Client<CountUntil>::SharedPtr count_until_client_;
    rclcpp::TimerBase::SharedPtr timer_;
    CountUntilGoalHandle::SharedPtr goal_handle_;
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilClientNode>(); 
    node->send_goal(15,1);
    node->send_goal(10,1);
    node->send_goal(5,1);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}